src/storage/s3/storage.c

```diff
@@ -171,6 +175,13 @@ storageS3Auth(
         if (this->securityToken != NULL)
             httpHeaderPut(httpHeader, S3_HEADER_TOKEN_STR, this->securityToken);
 
+        // Set KMS key header if this is a write request
+        if (this->kmsKeyId != NULL && !strCmp(verb, HTTP_VERB_PUT_STR))
+        {
```

Comment:
This doesn't look right. I think you mean `strCmp(verb, HTTP_VERB_PUT_STR)` but it seems at least POST would also be required. Sees like this should be implemented in the `StorageWriteS3` object instead where it is more obvious where it should be applied.

Reply:
I had originally included POST as well, but that's refused by aws with an "header not applicable here" error.

I don't understand the "you mean strCmp" part, that's exactly what I included there?

During development, I had tried to put the logic into StorageWriteS3, but failed since I couldn't access neither the kmskey option nor the header object there. I'll give it another try, but possibly I'll need help in sorting out the object-oriented C code there.

Reply:
The two sites in s3/write.c that use PUT are these:
```
storageWriteS3PartAsync(StorageWriteS3 *this)
{
...
            this->request = storageS3RequestAsyncP(
                this->storage, HTTP_VERB_PUT_STR, this->interface.name, .query = query, .content = this->partBuffer);
```
and
```
storageWriteS3Close(THIS_VOID)
{
...
            // If a multi-part upload was started we need to finish that way
            if (this->uploadId != NULL)
...
            // Else upload all the data in a single put
            else
                storageS3RequestP(this->storage, HTTP_VERB_PUT_STR, this->interface.name, .content = this->partBuffer);
```
The first one looks easy to modify since it has a "query" member, but the second one is weird as it's starting a new request on close.

To be honest, since this is about encryption, I would prefer to have the header added in the place where it's now, since that seems to be the function that all HTTP requests pass through. If we instead sprinkle the header additions in other places, we might miss one (or some future addition might neglect it). Centralizing that seems safer for security to me, as encryption is optional and AWS S3 does not raise an error if the header is missing.

Reply:
> The first one looks easy to modify since it has a "query" member, 

The first PUT is an UploadPart (https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPart.html) and does not require the key because it was already provided in the POST for CreateMultipartUpload (https://docs.aws.amazon.com/AmazonS3/latest/API/API_CreateMultipartUpload.html).

> but the second one is weird as it's starting a new request on close.

That second PUT writes any files that are < 5MB all at once and it requires the key (https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html).

So the verb is not enough to know whether this key is needed or not. Also, I think it should only be added to GET object and not GET list, DELETE, etc.

The easiest thing to do here I think is add a new param to `storageS3RequestP()` by adding a member to the `StorageS3RequestParam` struct:
```
bool sseKms; // ...
```
Then you just need to set it in the `storageS3RequestP()` calls where needed like `storageS3Request(..., sseKms = true)`. Add your header logic to `storageS3RequestAsync()` conditional on param.sseKms being set.

The way I read it you'll need to set sseKms at line 55 in s3/read.c and lines 126 and 251 in write.c, but I may be missing something.

Reply:
Hmmm, you'll also need to add `sseKms` to the `StorageS3RequestAsyncParam` struct and pass it through to `storageS3RequestAsync()` in `storageS3Request()`. You'd don't need async for this feature but that's where the request logic lives.

Reply:
The "may be missing something" part is why I think centralizing that is better.

Reply:
Agreed, and centralizing it what I'm describing above, but with the ability to turn it on only where needed.

Reply:
So, something like this: https://github.com/pgbackrest/pgbackrest/commit/1d45b2c25425cf6b145f9a0f577f755d1905a142

This will be a easier if you allow me to update your PR directly. See https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork.

Reply:
It's somewhat hidden in a wall of text on that page, but "allow edits from maintainers" works only for *user-owned* repositories, this fork is owned by the company account unfortunately.

Reply:
Thanks for picking this up.
As @tharinduamila-insta said in https://github.com/pgbackrest/pgbackrest/commit/1d45b2c25425cf6b145f9a0f577f755d1905a142#r60776961 , GET requests don't have the headers, so only the two callers in write.c should remain.
(I unfortunately deleted parts of my test S3 account, and I fail to recreate it, so my ability to test this is somewhat limited at the moment, but AWS is so kind to still spit out "no key allowed here" error messages.)